src/ray/core_worker/core_worker.h

```diff
@@ -1952,5 +1976,19 @@ class CoreWorker {
 
   /// Callback to free an RDT object when it is out of scope.
   std::function<void(const ObjectID &)> free_actor_object_callback_;
+
+  // Shutdown synchronization primitives
+  mutable absl::Mutex connected_mutex_;
+  bool connected_ ABSL_GUARDED_BY(connected_mutex_) = true;
+
+  mutable absl::Mutex shutdown_state_mutex_;
+  ray::core::ShutdownState shutdown_state_ ABSL_GUARDED_BY(shutdown_state_mutex_) =
+      ray::core::ShutdownState::kRunning;
```

Comment:
more mutexes is scary... more opportunities for deadlock. need to ensure we have an explicit total global order for acquiring these and document it.